lib/src/front_end/ast_node_visitor.dart

```diff
@@ -1336,19 +1336,70 @@ class AstNodeVisitor extends ThrowingAstVisitor<Piece> with PieceFactory {
 
   @override
   Piece visitRecordTypeAnnotation(RecordTypeAnnotation node) {
-    throw UnimplementedError();
+    var namedFields = node.namedFields;
+    var positionalFields = node.positionalFields;
+
+    // Single positional record types always have a trailing comma.
+    var listStyle = positionalFields.length == 1 && namedFields == null
+        ? const ListStyle(commas: Commas.alwaysTrailing)
+        : const ListStyle(commas: Commas.trailing);
+    var builder = DelimitedListBuilder(this, listStyle);
+
+    // If all parameters are optional, put the `{` right after `(`.
+    if (positionalFields.isEmpty && namedFields != null) {
+      builder.leftBracket(
+        node.leftParenthesis,
+        delimiter: namedFields.leftBracket,
+      );
+    } else {
+      builder.leftBracket(node.leftParenthesis);
+    }
+
+    for (var positionalField in positionalFields) {
+      builder.visit(positionalField);
+    }
+
+    Token? rightDelimiter;
+    if (namedFields != null) {
+      // If this is the first optional parameter, put the delimiter before it.
+      if (positionalFields.isNotEmpty) {
+        builder.leftDelimiter(namedFields.leftBracket);
+      }
+      for (var namedField in namedFields.fields) {
+        builder.visit(namedField);
+      }
+      rightDelimiter = namedFields.rightBracket;
+    }
+
+    builder.rightBracket(node.rightParenthesis, delimiter: rightDelimiter);
+    return buildPiece((b) {
+      b.add(builder.build());
+      if (node.question case var question?) {
+        b.add(tokenPiece(question));
```

Comment:
```dart
b.token(question);
```

AdjacentBuilder has helper functions for `token()` and `visit()` since adding a token or visited AST node is so common.

In fact, you can even just do:

```dart
b.token(node.question);
```

And get rid of the surrounding `if` because `AdjacentBuilder.token()` silently does nothing if you pass it `null`, since optional tokens are so common.

Reply:
Amazing. Will do and will remember for next time.